### PR TITLE
Update constraints for zarr and pyqt

### LINKED ISSUE #121: `pyproject.toml` was updated (2024-03-14)
The file `pyproject.toml` was modified in `napari/napari`. This might mean the conda-recipe in
`napari/packaging` needs adjusting! Please check!

Calculated diff:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index ab9af2d7..c205b777 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -121,6 +121,7 @@ all = [
 optional = [
     "triangle ; platform_machine != 'arm64'",
     "numba>=0.57.1",
+    "zarr>=2.12.0", # needed by `builtins` (dask.array.from_zarr) to open zarr
 ]
 testing = [
     "babel>=2.9.0",
```

cc @jaimergp

### LINKED ISSUE #123: `pyproject.toml` was updated (2024-03-21)
The file `pyproject.toml` was modified in `napari/napari`. This might mean the conda-recipe in
`napari/packaging` needs adjusting! Please check!

Calculated diff:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index ff56ccb9..0cf47605 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -101,6 +101,7 @@ pyside6_experimental = [
 ]
 pyqt6_experimental = [
     "PyQt6",
+    "PyQt6 != 6.6.1 ; platform_system == 'Darwin'"
 ]
 pyside = [
     "napari[pyside2]"
```

cc @jaimergp

### LINKED ISSUE #126: `pyproject.toml` was updated (2024-03-25)
The file [`pyproject.toml`](https://github.com/napari/napari/blob/main/pyproject.toml) (see [history](https://github.com/napari/napari/commits/main/pyproject.toml)) was modified in `napari/napari`. This might mean the conda-recipe in `napari/packaging` needs adjusting! Please check!

Calculated diff:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index bb034732..9e3dea94 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -66,7 +66,7 @@ dependencies = [
     "scipy>=1.4.1 ; python_version < '3.9'",
     "scipy>=1.5.4 ; python_version >= '3.9'",
     "superqt>=0.5.0",
-    "tifffile>=2020.2.16",
+    "tifffile>=2022.4.8",
     "toolz>=0.10.0",
     "tqdm>=4.56.0",
     "typing_extensions>=4.2.0",
@@ -565,7 +565,6 @@ module = [
     "napari.layers.shapes._shapes_models.line",
     "napari.layers.shapes._shapes_models.rectangle",
     "napari.layers.shapes._shapes_models.shape",
-    "napari.layers.vectors._vector_utils",
     "napari.resources._icons",
     "napari.utils.color",
     "napari.utils.events.containers._dict",
@@ -622,7 +621,6 @@ module = [
     "napari.layers.surface._surface_key_bindings",
     "napari.layers.tracks._tracks_key_bindings",
     "napari.layers.utils._slice_input",
-    "napari.layers.vectors._vectors_key_bindings",
     "napari.utils._register",
     "napari.utils.colormaps.categorical_colormap",
     "napari.utils.colormaps.standardize_color",
@@ -633,7 +631,8 @@ module = [
     "napari.utils.translations",
     "napari.utils.tree.node",
     "napari.viewer",
-    "napari.layers.shapes._shape_list"
+    "napari.layers.shapes._shape_list",
+    "napari.layers.vectors.vectors",
 ]
 disallow_incomplete_defs = false
 disallow_untyped_calls = false
@@ -646,7 +645,6 @@ module = [
     "napari._vispy.visuals.axes",
     "napari.layers.labels._labels_mouse_bindings",
     "napari.layers.utils.color_manager_utils",
-    "napari.layers.vectors._slice",
     "napari.utils.colormaps.vendored._cm",
     "napari.utils.colormaps.vendored.cm",
     "napari.utils.status_messages",
@@ -686,7 +684,6 @@ module = [
     "napari.layers.surface.surface",
     "napari.layers.tracks.tracks",
     "napari.layers.utils.layer_utils",
-    "napari.layers.vectors.vectors",
     "napari.utils._dtype",
     "napari.utils.colormaps.colormap_utils",
     "napari.utils.misc",
```

cc @jaimergp

### LINKED ISSUE #128: `pyproject.toml` was updated (2024-03-26)
The file [`pyproject.toml`](https://github.com/napari/napari/blob/main/pyproject.toml) (see [history](https://github.com/napari/napari/commits/main/pyproject.toml)) was modified in `napari/napari`. This might mean the conda-recipe in `napari/packaging` needs adjusting! Please check!

Calculated diff:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 9e3dea94..52578e9e 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,7 +1,7 @@
 [build-system]
 requires = [
-  "setuptools >= 61.2",
-  "setuptools_scm[toml]>=3.4"
+  "setuptools >= 69",
+  "setuptools_scm[toml]>=8"
 ]
 build-backend = "setuptools.build_meta"
 
@@ -20,10 +20,10 @@ classifiers = [
     "Programming Language :: C",
     "Programming Language :: Python",
     "Programming Language :: Python :: 3 :: Only",
-    "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
     "Topic :: Scientific/Engineering",
     "Topic :: Scientific/Engineering :: Visualization",
     "Topic :: Scientific/Engineering :: Information Analysis",
@@ -34,7 +34,7 @@ classifiers = [
     "Operating System :: Unix",
     "Operating System :: MacOS",
 ]
-requires-python = ">=3.8"
+requires-python = ">=3.9"
 dependencies = [
     "appdirs>=1.4.4",
     "app-model>=0.2.2,<0.3.0",
@@ -51,8 +51,7 @@ dependencies = [
     "npe2>=0.7.2",
     "numpy>=1.22.2,<2",
     "numpydoc>=0.9.2",
-    "pandas>=1.1.0 ; python_version < '3.9'",
-    "pandas>=1.3.0 ; python_version >= '3.9'",
+    "pandas>=1.3.0",
     "Pillow>=9.0",
     "pint>=0.17",
     "psutil>=5.0",
@@ -63,8 +62,7 @@ dependencies = [
     "PyYAML>=5.1",
     "qtpy>=1.10.0",
     "scikit-image[data]>=0.19.1",
-    "scipy>=1.4.1 ; python_version < '3.9'",
-    "scipy>=1.5.4 ; python_version >= '3.9'",
+    "scipy>=1.5.4",
     "superqt>=0.5.0",
     "tifffile>=2022.4.8",
     "toolz>=0.10.0",
@@ -93,11 +91,10 @@ Documentation = "https://napari.org"
 
 [project.optional-dependencies]
 pyside2 = [
-    "PySide2>=5.13.2,!=5.15.0 ; python_version != '3.8'",
-    "PySide2>=5.14.2,!=5.15.0 ; python_version == '3.8'",
+    "PySide2>=5.13.2,!=5.15.0 ; python_version < '3.12'",
 ]
 pyside6_experimental = [
-    "PySide6",
+    "PySide6 < 6.5 ; python_version < '3.12'"
 ]
 pyqt6_experimental = [
     "PyQt6",
@@ -201,7 +198,7 @@ napari_builtins = [
 write_to = "napari/_version.py"
 
 [tool.black]
-target-version = ['py38', 'py39', 'py310']
+target-version = ['py39', 'py310', 'py311', 'py312']
 skip-string-normalization = true
 line-length = 79
 exclude = '''
@@ -269,7 +266,7 @@ exclude = [
     "*_vendor*",
 ]
 
-target-version = "py38"
+target-version = "py39"
 fix = true
 
 [tool.ruff.lint]
@@ -299,7 +296,7 @@ select = [
     "RUF", # ruff specyfic rules
 ]
 ignore = [
-    "E501", "UP006", "UP007", "TCH001", "TCH002", "TCH003",
+    "E501", "TCH001", "TCH002", "TCH003",
     "A003", # flake8-builtins - we have class attributes violating these rule
     "COM812", # flake8-commas - we don't like adding comma on single line of arguments
     "SIM117", # flake8-simplify - we some of merged with statements are not looking great with black, reanble after drop python 3.9
@@ -321,6 +318,9 @@ ignore = [
 "**/vendored/**" = ["TID"]
 "napari/benchmarks/**" = ["RUF012", "TID252"]
 
+[tool.ruff.lint.pyupgrade]
+keep-runtime-typing = true
+
 [tool.ruff.lint.flake8-quotes]
 docstring-quotes = "double"
 inline-quotes = "single"
```

cc @jaimergp

### LINKED ISSUE #133: `pyproject.toml` was updated (2024-04-14)
The file [`pyproject.toml`](https://github.com/napari/napari/blob/main/pyproject.toml) (see [history](https://github.com/napari/napari/commits/main/pyproject.toml)) was modified in `napari/napari`. This might mean the conda-recipe in `napari/packaging` needs adjusting! Please check!

Calculated diff:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index b135e7a1..a99f9e08 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -37,7 +37,7 @@ classifiers = [
 requires-python = ">=3.9"
 dependencies = [
     "appdirs>=1.4.4",
-    "app-model>=0.2.2,<0.3.0",
+    "app-model>=0.2.5,<0.3.0",
     "cachey>=0.2.1",
     "certifi>=2018.1.18",
     "dask[array]>=2021.10.0",
@@ -48,7 +48,7 @@ dependencies = [
     "napari-console>=0.0.9",
     "napari-plugin-engine>=0.1.9",
     "napari-svg>=0.1.8",
-    "npe2>=0.7.2",
+    "npe2>=0.7.4",
     "numpy>=1.22.2,<2",
     "numpydoc>=0.9.2",
     "pandas>=1.3.0",
@@ -135,7 +135,7 @@ testing = [
     "pyautogui>=0.9.54",
     "pytest-qt>=4.3.1",
     "pytest-pretty>=1.1.0",
-    "pytest>=8.0.0",
+    "pytest>=8.1.0",
     "tensorstore>=0.1.13",
     "virtualenv>=20.17",
     "xarray>=0.16.2",
@@ -244,7 +244,6 @@ ignore = [
 
 [tool.ruff]
 line-length = 79
-
 exclude = [
     ".bzr",
     ".direnv",
@@ -308,7 +307,6 @@ ignore = [
     "PYI015", # it produces bad looking files (@jni opinion)
 ]
 
-
 [tool.ruff.lint.per-file-ignores]
 "napari/_vispy/__init__.py" = ["E402"]
 "**/_tests/*.py" = ["B011", "INP001", "TRY301", "B018", "RUF012"]
@@ -344,10 +342,12 @@ xarray = "xr"
 [tool.pytest.ini_options]
 # These follow standard library warnings filters syntax.  See more here:
 # https://docs.python.org/3/library/warnings.html#describing-warning-filters
-addopts = "--maxfail=5 --durations=10 -rXxs"
+addopts = ["--maxfail=5",  "--durations=10",  "-ra", "--strict-markers", "--strict-config"]
 console_output_style = "count"
-pystack_threshold=60
-pystack_args="--native-all"
+minversion = "8"
+# log_cli_level = "INFO"
+xfail_strict = true
+testpaths = ["napari", "napari_builtins"]
 
 
 # NOTE: only put things that will never change in here.
@@ -388,7 +388,7 @@ files = "napari"
 exclude = "napari/utils/events/_tests/test_evented_model\\.py"
 plugins =  "numpy.typing.mypy_plugin, pydantic.mypy"
 ignore_missing_imports = true
-show_error_codes = true
+hide_error_codes = false
 warn_redundant_casts = true
 disallow_incomplete_defs = true
 disallow_untyped_calls = true
```

cc @jaimergp

### PR DESCRIPTION
**Pull request by jaimergp**
- Closes #121
- Closes #123
- Closes #126
- Closes #128
- Closes #133

### CODE REVIEW

File: conda-recipe/meta.yaml [hunk 1]
```diff
@@ -79,12 +79,12 @@ outputs:
...
         - fsspec
         - pooch >=1.3.0
-        - zarr
+        - zarr >=2.12.0
 
       run_constrained:
         - pyside2 >=5.14.2,!=5.15.0
-        - pyqt >=5.12.3,!=5.15.0
         - pyside6 <6.3.2
+        - pyqt >=5.12.3,!=5.15.0,!=6.6.1
```
Comment (jaimergp) [inline conda-recipe/meta.yaml:87]: Note [pyqt6 doesn't exist yet on conda-forge](https://github.com/conda-forge/pyqt-feedstock/pull/136), but for completeness, I'm adding it here.

File: conda-recipe/meta.yaml [hunk 2]
```diff
@@ -79,12 +79,12 @@ outputs:
         # additional dependencies for convenience in conda-forge
         - fsspec
         - pooch >=1.3.0
-        - zarr
+        - zarr >=2.12.0
 
       run_constrained:
         - pyside2 >=5.14.2,!=5.15.0
-        - pyqt >=5.12.3,!=5.15.0
         - pyside6 <6.3.2
```
Comment (jaimergp) [inline conda-recipe/meta.yaml:86]: I see that this constraint is no longer mentioned in pyproject.toml, but [there are several entries in the `denylist`](https://github.com/psobolewskiPhD/napari/blob/2e5a7fe83ec0ad1a626dffbd8255f9f286091ce7/resources/constraints/version_denylist.txt#L3C9-L3C95)... Are we supposed to track those in the conda recipe too? I'm not sure what the role of that constraint file is 🤔 

Do you have more details @psobolewskiPhD?

Comment (psobolewskiPhD) [reply conda-recipe/meta.yaml]: Hmm, I stil see it in pyproject.toml:
https://github.com/napari/napari/blob/c864d6e41a900721517c1cf0654f3b02ec3fcb01/pyproject.toml#L106-L108

Or are you referring to pyside6?
That one is a big problem no combo of napari, python, and pyside6 works 100% properly.
The deny list is used to define what combination can pass on CI, but it has glitches in regular usage including flashing windows and seg faults.
It's very much experimental and not recommended.

Comment (jaimergp) [reply conda-recipe/meta.yaml]: Yes, I was referring to `pyside6`. Should we prevent its installation altogether then?  I can also update it to:

```
PySide6 <6.5
```

so it tracks `pyside6_experimental`.

Comment (psobolewskiPhD) [reply conda-recipe/meta.yaml]: Right now it look like <6.5 or >= 6.7 might actually result in run-able napari.
With all the other caveats.
I dunno about preventing it all together, maybe people want to test?

Comment (jaimergp) [reply conda-recipe/meta.yaml]: Ok, let's add `<6.5,>=6.7` then. Thanks!

**Pull request merged**